docs/programming-oak.md

```diff
@@ -0,0 +1,305 @@
+configuration (described in the next section) easier:
+
+[embedmd]:# (../examples/rustfmt/client/rustfmt.cc C++ /.*Connect to the Oak Manager/ /FLAGS_manager_address.*;/)
+```C++
+  // Connect to the Oak Manager.
+  std::unique_ptr<oak::ManagerClient> manager_client =
+      absl::make_unique<oak::ManagerClient>(grpc::CreateChannel(
+          absl::GetFlag(FLAGS_manager_address), grpc::InsecureChannelCredentials()));
+```
+
+### Starting the Oak Application
+
+To use the `CreateApplication` method, the client needs to specify the
+*application configuration*; this describes the Nodes that make up the
+application, and their connectivity.  However, for the simple case of a one-Node
+application the `oak::ManagerClient` helper takes care of this; the client
+just need to provide the code that is going to be run in the Node, as a Wasm
+module:
+
+[embedmd]:# (../examples/rustfmt/client/rustfmt.cc C++ /.*FLAGS_module/ /}/)
+```C++
+  std::string module_bytes = oak::utils::read_file(absl::GetFlag(FLAGS_module));
+  std::unique_ptr<oak::CreateApplicationResponse> create_application_response =
+      manager_client->CreateApplication(module_bytes);
+  if (create_application_response == nullptr) {
+    LOG(QFATAL) << "Failed to create application";
+  }
+```
+
+There's also a variant of `CreateApplication` that allows the client code to
+specify a storage provider to connect to, and to control whether logging is
+enabled:
+
+[embedmd]:# (../examples/hello_world/client/hello_world.cc C++ /.*FLAGS_module/ /}/)
+```C++
+  std::string module_bytes = oak::utils::read_file(absl::GetFlag(FLAGS_module));
+  std::unique_ptr<oak::CreateApplicationResponse> create_application_response =
+      manager_client->CreateApplication(module_bytes, /* logging= */ true,
+                                        absl::GetFlag(FLAGS_storage_address));
+  if (create_application_response == nullptr) {
+    LOG(QFATAL) << "Failed to create application";
+  }
+```
+
+The Oak Manager will launch an [Oak VM](concepts.md#oak-vm) inside the enclave,
+and this VM will check the provided Wasm module(s) and application
+configuration.  Assuming everything is correct (e.g. the Nodes all have an
+`oak_main` entrypoint and only expect to find the Oak [host
+functions](abi.md#host-functions)), the Oak VM opens up a port of its own and
+the `CreateApplication` returns this to the client.
+
+The client can now connect to this separate gRPC service, and send
+(Node-specific) gRPC requests to it, over a channel that has end-to-end
+encryption into the enclave:
+
+[embedmd]:# (../examples/rustfmt/client/rustfmt.cc C++ /.*InitializeAssertionAuthorities/ /CredentialsOption/)
+```C++
+  oak::ApplicationClient::InitializeAssertionAuthorities();
+
+  // Connect to the newly created Oak Application.
+  auto stub = FormatService::NewStub(grpc::CreateChannel(
+      addr.str(), asylo::EnclaveChannelCredentials(asylo::BidirectionalNullCredentialsOption
+```
+
+## Request Processing Path
```

Comment:
This heading wasn't clear to me at first, clarify that it's gRPC request processing?

Reply:
Done.